### PR TITLE
[PLGN-421] Refactor joe sandbox | Upgrade SDK | Update jbxapi dependency

### PR DESCRIPTION
**Pull request by cmcnally-r7**
### Description

Describe the proposed changes:

  - Update SDK
  - Update `jbxapi` dependency
  - Both == resolved connection issues

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``rapid7/insightconnect-python-3-38-slim-plugin:{sdk-version-num}`` and ``rapid7/insightconnect-python-3-38-plugin:{sdk-version-num}``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``insight-plugin validate`` which calls ``icon_validate`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `insight-plugin samples`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `insight-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `insight-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `insight-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `insight-plugin run --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `insight-plugin validate` and make sure everything passes
2. Run the assessment tool: `insight-plugin run -A`. For single action validation: `insight-plugin run tests/{file}.json -A`
3. Copy (`insight-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

File: plugins/joe_sandbox/unit_test/mock.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        return MockResponse("get_submitted_info", status_code)
+    if url == "https://example.com/v2/analysis/list":
+        return MockResponse("list_analysis", status_code)
+    if url == "https://example.com/v2/server/lia_countries":
+        return MockResponse("list_countries", status_code)
+    if url == "https://example.com/v2/server/systems":
+        return MockResponse("list_systems", status_code)
+    if url == "https://example.com/v2/analysis/search":
+        return MockResponse("search_analysis", status_code)
+    # if url == "https://example.com/v2/submission/new":
```
Comment (dlaverty-r7) [inline plugins/joe_sandbox/unit_test/mock.py:70]: Can these be removed if they are unused?

File: plugins/joe_sandbox/help.md [hunk 2]
```diff
@@ -5,607 +5,660 @@
...
+*This plugin does not contain any triggers.*
+### Tasks
+  
+*This plugin does not contain any tasks.*
+
+### Custom Types
+  
+**system**
+
+|Name|Type|Default|Required|Description|Example|
```
Comment (dlaverty-r7) [inline plugins/joe_sandbox/help.md:582]: Can examples be added for the custom type?

**Review by dlaverty-r7 — APPROVED** [review]
Generally lgtm, as discussed, add back in the removed action for now if its still functions ok.

File: plugins/joe_sandbox/icon_joe_sandbox/connection/connection.py [hunk 3]
```diff
@@ -1,20 +1,20 @@
...
+class Connection(insightconnect_plugin_runtime.Connection):
     def __init__(self):
         super(self.__class__, self).__init__(input=ConnectionSchema())
 
     def connect(self, params):
         self.logger.info("Connect: Connecting...")
 
-        api_key = params.get("api_key").get("secretKey")
-        api_url = params.get("url")
+        api_key = params.get(Input.API_KEY).get("secretKey")
```
Comment (igorski-r7) [inline plugins/joe_sandbox/icon_joe_sandbox/connection/connection.py:16]:
```suggestion
        api_key = params.get(Input.API_KEY, {}).get("secretKey")
```

File: plugins/joe_sandbox/icon_joe_sandbox/connection/connection.py [hunk 4]
```diff
@@ -40,8 +43,8 @@ def test(self):
             raise ConnectionTestException(preset=ConnectionTestException.Preset.SERVER_ERROR)
         except jbxapi.PermissionError:
             raise ConnectionTestException(preset=ConnectionTestException.Preset.UNAUTHORIZED)
-        except jbxapi.ApiError as e:
+        except jbxapi.ApiError as error:
             raise ConnectionTestException(
-                cause="An error occurred: " + str(e),
+                cause="An error occurred: " + str(error),
```
Comment (igorski-r7) [inline plugins/joe_sandbox/icon_joe_sandbox/connection/connection.py:48]:
```suggestion
                cause=f"An error occurred: {str(error)}",
```

File: plugins/joe_sandbox/plugin.spec.yaml
Comment (igorski-r7) [inline plugins/joe_sandbox/plugin.spec.yaml:1]: Please add new line at the end of file

File: plugins/joe_sandbox/requirements.txt [hunk 5]
```diff
@@ -1,4 +1,4 @@
 # List third-party dependencies here, separated by newlines. 
 # All dependencies must be version-pinned, eg. requests==1.2.0
 # See: https://pip.pypa.io/en/stable/user_guide/#requirements-files
-jbxapi==2.9.5
+jbxapi==3.21.0
```
Comment (igorski-r7) [inline plugins/joe_sandbox/requirements.txt:4]: Could you please update to the latest one `3.22.0` or there is a cause of having `3.21.0`? 😄

File: plugins/joe_sandbox/unit_test/__init__.py
Comment (igorski-r7) [inline plugins/joe_sandbox/unit_test/__init__.py:1]: Please add new line at the end of file

File: plugins/joe_sandbox/unit_test/mock.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    #     return MockResponse("submit_cookbook", status_code)
+    # return MockResponse("submit_sample", status_code)
+    # return MockResponse("submit_sample_url", status_code)
+    # return MockResponse("submit_url", status_code)
```
Comment (igorski-r7) [inline plugins/joe_sandbox/unit_test/mock.py:74]: What about those commented lines?

File: plugins/joe_sandbox/unit_test/mock.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        return action
+
+
+class MockResponse:
+    def __init__(self, filename: str, status_code: 200) -> None:
+        self.filename = filename
+        self.status_code = status_code
+        self.text = json.dumps(self.json())
+        self.ok = True
+        # self.iter_content = json.dumps(self.json)
```
Comment (igorski-r7) [inline plugins/joe_sandbox/unit_test/mock.py:32]: Should be leave it?

File: plugins/joe_sandbox/unit_test/test_download_analysis.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+#         self.params = {Input.WEBID: "abc", Input.TYPE: "html", Input.RUN: 0}
+#
+#     @patch("requests.request", side_effect=mock_request_200)
+#     def test_download_analysis(self, mock_get):
+#         mocked_request(mock_get)
+#         response = self.action.run()
+#
+#         expected = {Output.RESOURCE_NAME: "name", Output.RESOURCE_CONTENT: "content"}
+#         validate(response, self.action.output.schema)
+#         self.assertEqual(response, expected)
```
Comment (igorski-r7) [inline plugins/joe_sandbox/unit_test/test_download_analysis.py:27]: What about those tests? 😄

File: plugins/joe_sandbox/unit_test/test_list_analyses.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+from mock import Util, mock_request_200, mocked_request
+
+
+class TestListAnalyses(TestCase):
+    @patch("requests.request", side_effect=mock_request_200)
+    def setUp(self, mock_client) -> None:
+        self.action = Util.default_connector(ListAnalyses())
+
+    @patch("requests.request", side_effect=mock_request_200)
+    def test_list_analyses(self, mock_get):
```
Comment (igorski-r7) [inline plugins/joe_sandbox/unit_test/test_list_analyses.py:20]:
```suggestion
    def test_list_analyses(self, mock_get: MagicMock) -> None:
```

File: plugins/joe_sandbox/requirements.txt (hunk 5 above)
Comment (cmcnally-r7) [reply plugins/joe_sandbox/requirements.txt:4]: I followed along with the customers update but that was back in November so I can update this now :)

File: plugins/joe_sandbox/icon_joe_sandbox/actions/list_keyboard_layouts/action.py
Comment (cmcnally-r7) [inline plugins/joe_sandbox/icon_joe_sandbox/actions/list_keyboard_layouts/action.py:1]: Regarding this action, the original method `get_keyboard_layouts()` was changed in `v3.0.0` to now be `server_languages_and_locales()`.
https://github.com/joesecurity/jbxapi/blob/master/CHANGES.md#version-300

**Review by igorski-r7 — APPROVED** [review]

**Pull request merged**